virtool/administrators/data.py

```diff
@@ -1,186 +1,258 @@
     ResourceNotFoundError,
+    ResourceConflictError,
+)
+from virtool.data.piece import DataLayerPiece
+from virtool.errors import DatabaseError
+from virtool.groups.db import lookup_groups_minimal_by_id, lookup_group_minimal_by_id
+from virtool.users.db import (
+    fetch_complete_user,
+    compose_primary_group_update,
+    update_keys,
+    compose_groups_update,
 )
 
 if TYPE_CHECKING:
     from virtool.mongo.core import DB
 
 
-AVAILABLE_ROLES = [
-    {"id": role, "name": role.capitalize(), "description": role.__doc__}
-    for role in document_enum(AdministratorRole)
+PROJECTION = [
+    "_id",
+    "handle",
+    "administrator",
+    "force_reset",
+    "groups",
+    "last_password_change",
+    "permissions",
+    "primary_group",
+    "administrator_role",
+    "active",
+    "b2c",
+    "b2c_display_name",
+    "b2c_family_name",
+    "b2c_given_name",
+    "b2c_oid",
 ]
 
 
-class AdministratorsData:
-    def __init__(self, authorization_client: AuthorizationClient, db: "DB"):
+class AdministratorsData(DataLayerPiece):
+    def __init__(self, authorization_client: AuthorizationClient, mongo: "DB"):
         self._authorization_client = authorization_client
-        self._db = db
-
-    async def find(self) -> AdministratorSearch:
+        self._mongo = mongo
+
+    async def find_users(
+        self,
+        page: int,
+        per_page: int,
+        administrator: Optional[bool] = None,
+        term: Optional[str] = None,
+    ) -> UserSearchResult:
         """
-        List all administrators.
-
-        :return: a list of all administrators
+        Find users. Optionally filter by administrator status and/or search term.
 
+        :param page: the page number
+        :param per_page: the number of items per page
+        :param administrator: whether to filter by administrator status
+        :param term: a search term to filter by user handle
         """
-        admin_list = await self._authorization_client.list_administrators()
-
-        return AdministratorSearch(
-            **{
-                "items": [
-                    AdministratorMinimal(
-                        **{
-                            **await self._db.users.find_one(admin_tuple[0]),
-                            "role": AdministratorRole(admin_tuple[1]),
-                        }
-                    )
-                    for admin_tuple in admin_list
-                ],
-                "available_roles": AVAILABLE_ROLES,
-            }
+
+        admin_dict = dict(await self._authorization_client.list_administrators())
+
+        administrator_query = {}
+        if administrator is not None:
+            operator = "$in" if administrator else "$nin"
+            administrator_query = {"_id": {operator: list(admin_dict.keys())}}
+
+        term_query = compose_regex_query(term, ["handle"]) if term else {}
+
+        client_query = {
+            **administrator_query,
+            **term_query,
+        }
```

Comment:
Converting to `dict`, then calling `keys()` is indirect and not particularly efficient. It is also not necessary to access OpenFGA if `administrator is None` .

Not sure if the unary will work for `dict` keys, but my guess is that it will. Up to you if you want to do it.

```suggestion
        query = {}
        
        if administrator is not None:
            administrator_ids = [a[0] for a in await self._authorization_client.list_administrators()]            
            query["_id"]: {"$in" if administrator else "$nin": administrator_ids}
            
        if term:
            query.update(compose_regex_query(term, ["handle"]))
```

Reply:
Gonna split this into separate items for clarity:

 1. Accessing openFGA is required regardless of filtering, since the endpoint returns a user object with `administrator_role` attached. Unfortunately that info is not stored in mongo. If there is a large number of admins there may be an argument for aggregating the users to be sent, then requesting specific tuples for the users. Given the small number of admins and the relatively large number of users my suspicion is that it will be faster as is though.
 
 2. Your right the dict is indirect. That said, I converted it into a dict since the dict is O(c) for doing lookups when a attaching admin roles to the returned users. Though given how short the admin list is it might be possible that it would be faster to eat the O(n) of searching through the whole list when attaching the ids. I need to retain the `role:id` association from the tuple, so the list soln would have to be a list of tuples or similar.

Reply:
1. Right. I missed this in the GitHub diff view.
2. I see how the `dict` makes sense now. Keep it as is. Please come up with a better name than `admin_dict`. Names with types in them are a code smell (even though I have git blame on some of them...).